content/posts/2018-10-30_Docker-dev-and-prod-ready-setup-for-SPA-app.md

```diff
@@ -0,0 +1,435 @@
+
+* `version: 3.6` - Shows supported compose file version to `docker-compose`, so we'll know what syntax and features we could use.
+* `services:` - main compose purpose, registering services to run them all neither using a command for one by one like `docker run ...`
+* `client:` - registering our first and main service - `client` which will be our React app,
+* `env_file:` - pointing the `.env` file which should be used to get global environments values for our setup. This file helps us to storage variables for our app like exposed port, application environment, compose project name and much more, that we could easily change due to fact that they are in one place.
+
+Okay, now our **Docker** setup contains only base `docker-compose.yml`, so we need to add first override compose file for development environment: `docker-compose.dev.yml`
+
+```yml
+# docker-compose.dev.yml
+version: '3.6'
+
+services:
+  client:
+    image: node:carbon
+    volumes:
+      - ..:/opt/app:cached
+    ports:
+      - "${NODE_PORT}:${NODE_PORT}"
+    working_dir: /opt/app
+    command: "/bin/sh -c 'yarn install --frozen-lockfile --production=false; yarn start'"
+```
+
+Describe a bit more this little magic:
+
+* `image:` - here we're specifying that we want to use an image with predefined `node.js` in `carbon` (`8.x`) version. Based on this image, Docker will build a container for our application (`client`),
+* `volumes:` - this allows us to mount some data volume (we needn't copy data, because we mount data from our host to container) as `HOST:CONTAINER` path. So we're mounting `..` from the host (project root directory) to `/opt/app` in our container. We also add `:cached` to improving Mac's performance. More details about it [here](https://docs.docker.com/docker-for-mac/osxfs-caching/#cached).
+* `ports:` - here we assign port from our host to port in our container (`HOST:CONTAINER`), so we could access our container from outside e.g. `5000:5000` maps port `5000` from host to port `5000` in the container. We could also parametrise this values so we're getting them from env variables,
+* `working_dir:` - specifies root app directory, where `docker-compose` will be working,
+* `command:` - command which will run our container. We're passing to shell our commands (via `/bin/sh -c`), which firstly made installing dependencies with frozen lockfile (don't generate new `yarn.lockfile` and should download this same versions) and forced production flag to false (be sure, that `devDependencies` also will be installed).
+
+### <a name="default-env-vars"></a>Default environment variables
+
+For now, our Docker setup won't do anything useful for us.
+Also, if you'll try now to run it, compose will warn you about lack of image or build source.
+Firstly, let's add `.env` template file called `.env.dev.dist`!
+
+```apacheconf
+# .env.dev.dist
+COMPOSE_PROJECT_NAME=my-app-spa
```

Comment:
Można już używać myślników? Kiedyś chyba wszystko było wycinane poza literami i cyframi, ale może to ulepszyli?

Reply:
u mnie działa :smile: